### DISCUSSION TITLE
A question about function "BIO_new"

### DISCUSSION THREAD

**Discussion by Suphappy** [openssl/openssl#21699] [Q&A]
Hello, why is it that sometimes but not always when the function "BIO_new" is called, the variable "b64" will be NULL. How do I handle this situation? Do I need to iterate until I return a normal address? Looking forward to your reply!

The code is as follows:

BIO *b64 = BIO_new(BIO_f_base64());
assert(b64);

**Comment by nhorman** [comment]
From the man page:
https://www.openssl.org/docs/manmaster/man3/BIO_new.html
Under the return value section, BIO_new returns NULL if the call to allocate a new BIO from the heap fails.

Standard error handling applies here, and the cause of the error can be retrieved from the library via the ERR api:
https://www.openssl.org/docs/manmaster/man3/ERR_get_error.html
Looking at the implementation current reasons the call might fail include:
ERR_R_MALLOC_FAILURE - no enough heap to fulfill the request
ERR_R_INIT_FAIL - the underlying implementation encountered an error setting up the BIO - In the case of the BIO_f_base64 method, that also amounts to a heap allocation failure, but there may be other causes for other BIO_methods.

How to proceed with error handling beyond fetching the error cause is the purview of the applcation using openssl.  That may amount to retrying the call again, potentially freeing up ram first, simply having the application fail whatever operating it was attempting, or whatever else might be relevant to your application

 
hope that helps.

**Reply by Suphappy** [reply]
Thank you for your answer, but I have checked that the system has enough free memory available for allocation, why does the heap fail to allocate?

**Reply by nhorman** [reply]
As I noted above, you have to augment your code to query the ERR api, that will provide you with an error code to determine what has gone wrong.  Any number of things are possible.  If you're using an older opensll library there may be other error signals raised.  Its also possible you are still having memory constraint issue (even if the system has sufficient free memory, for instance if you are running in a systemd slice or container that limits the amount of memory your application is allowed to allocate)

**Comment by t8m** [comment]
> Thank you for your answer, but I have checked that the system has enough free memory available for allocation, why does the heap fail to allocate?

This might indicate some internal error in your application that is corrupting the heap.

**Comment by nhorman** [comment]
If thats the case, running your application under valgrind may well illuminate the problem as well
